### ISSUE TITLE
Debugging duplicate dependency name error is difficult

### ISSUE THREAD

**Issue by Susurrus** [rust-lang/cargo#3592]
I have a project that has a bunch of dependency sections for different targets and supports quite a few. When building I'm getting an error like:
```shell
$ cargo build --target=arm-linux-androideabi --verbose
error: failed to parse manifest at `/home/susurrus/Projects/serialport-rs/Cargo.toml`

Caused by:
  found duplicate dependency name ioctl-rs, but all dependencies must have a unique name
```

This error was added as #2023. But looking at my Cargo.toml I can't see why this error should be occurring because the dependencies section looks like:
```toml
[dependencies]
libc = "0.2.1"

[target.arm-linux-androideabi.dependencies]
ioctl-rs = { path = "../ioctl" }
termios = { path = "../termios" }

[target.aarch64-unknown-linux-gnu.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.arm-unknown-linux-gnueabi.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.arm-unknown-linux-gnueabihf.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.armv7-unknown-linux-gnueabihf.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.i686-apple-darwin.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.i686-unknown-linux-gnu.dependencies]
ioctl-rs = "0.1.5"
libudev = "0.2.0"
termios = "0.2.2"

[target.mips-unknown-linux-gnu.dependencies]
ioctl-rs = "0.1.5"
libudev = "0.2.0"
termios = "0.2.2"

[target.mipsel-unknown-linux-gnu.dependencies]
ioctl-rs = "0.1.5"
libudev = "0.2.0"
termios = "0.2.2"

[target.powerpc-unknown-linux-gnu.dependencies]
ioctl-rs = "0.1.5"
libudev = "0.2.0"
termios = "0.2.2"

[target.x86_64-apple-darwin.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.x86_64-unknown-dragonfly.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.x86_64-unknown-freebsd.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.x86_64-unknown-linux-gnu.dependencies]
ioctl-rs = "0.1.5"
libudev = "0.2.0"
termios = "0.2.2"

[target.x86_64-unknown-openbsd.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"

[target.i686-pc-windows-msvc.dependencies]
winreg = "0.3"

[target.x86_64-pc-windows-msvc.dependencies]
winreg = "0.3"

[target.i686-pc-windows-gnu.dependencies]
winreg = "0.3"

[target.x86_64-pc-windows-gnu.dependencies]
winreg = "0.3"
```

So when building with cargo targetting `arm-linux-androideabi` I'd only expect three dependencies to be found: `libc`, `ioctl-rs`, and `termios`. But this doesn't appear to be the case and I have no idea why.

If this is not a parsing bug (or error on my end) I'd like to improve the error message added in #2023 to include what sections the duplicate dependencies are in so this can be more easily resolved.

**Comment by alexcrichton** [comment]
While this should likely be fixed regardless, I wonder if you'd want to start using the newer syntax?

```toml
[target.'cfg(unix)'.dependencies]
ioctl-rs = "0.1.3"

[target.'cfg(windows)'.dependencies]
winreg = "0.3"
```

etc...

**Comment by Susurrus** [comment]
I'll definitely be going that direction, I inherited this `Cargo.toml` from another project and never bothered to modernize it, but it's on the task list.

**Comment by Susurrus** [comment]
So interesting, here's a minimal case that will cause the failure:
```toml
[target.arm-linux-androideabi.dependencies]
ioctl = { path = "../ioctl" }
termios = { path = "../termios-rs" }

[target.arm-unknown-linux-gnueabi.dependencies]
ioctl-rs = "0.1.5"
termios = "0.2.2"
```

Now if I use that and build it on x86_64 linux using `cargo build --target=arm-linux-androideabi` I end up with the error.

Obviously there should be no error here as the dependencies are for different targets.

**Comment by Susurrus** [comment]
Looking at the relevant commit, it's clear that the target isn't checked. I think the correct behavior is to only check overlapping dependencies and to restrict this to just for the current build target.

**Comment by alexcrichton** [comment]
Ah yeah that's definitely intended to be an error. The "termios" crate there is coming from two different locations, which Cargo isn't intended to support.

**Comment by alexcrichton** [comment]
Although this seems reasonable to support I think in an abstract sense, so I'd be cool w/ a PR either way.

**Comment by Susurrus** [comment]
So all dependencies across the entire project must be from the same source (path, crates.io, or repository)? I'll try to whip up a PR with just improved text for this error as it's not clear that the "dependency name" in the error message really measure the code source for the dependency.

**Comment by alexcrichton** [comment]
PR looks good t ome!

**Comment by Susurrus** [comment]
This issue is resolved with the merging of #3593.
